src/test/java/com/stripe/param/common/HasExtraParamsSerializationTest.java

```diff
@@ -0,0 +1,116 @@
+package com.stripe.param.common;
+
+
+import static junit.framework.TestCase.assertEquals;
+
+import com.stripe.net.ApiRequestParams;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Test;
+
+public class HasExtraParamsSerializationTest {
+
+  enum ParamCode implements ApiRequestParams.EnumParam {
+    ENUM_FOO("enum_foo"),
+    ENUM_BAR("enum_bar"),;
+
+    private final String value;
+
+    ParamCode(String value) {
+      this.value = value;
+    }
+
+    @Override
+    public String getValue() {
+      return this.value;
+    }
+  }
+
+  private static class ModelHasExtraParams extends ApiRequestParams implements HasExtraParams {
+    private String stringValue;
+    private EnumParam enumValue;
+
+    public ModelHasExtraParams(EnumParam enumValue) {
+      this.stringValue = "foo";
+      this.enumValue = enumValue;
+      this.extraParams = new HashMap<>();
+      this.extraParams.put("hello", "world");
+    }
+
+    @Override
+    public Map<String, String> getExtraParam() {
+      return this.extraParams;
+    }
+
+    static Map<String, Object> expectedParams(String enumString) {
+      Map<String, Object> paramMap = new HashMap<>();
+      paramMap.put("string_value", "foo");
+      paramMap.put("enum_value", enumString);
+      paramMap.put("hello", "world");
```

Comment:
Make sure that it properly encodes non alphanumeric characters like `payment_intent_data[token]` for example

Reply:
Thank you @remi-stripe for the reminder!
Actually my preferred implementation here https://github.com/stripe/stripe-java/pull/752 will not require users to put in form encoding key.. The API looks like this:

```
    Map<String, Object> paymentIntentData = ImmutableMap.of(
        "application_fee_amount", 99L,
        "token", "src_123"
    );
    SessionCreateParams sessionCreateParams = SessionCreateParams.builder()
        .putExtraParam("payment_intent_data", paymentIntentData)
        .build();

    assertEquals(
        // nested form encoding from "extra params"
        "payment_intent_data[application_fee_amount]=99&payment_intent_data[token]=src_123",
        LiveStripeResponseGetter.createQuery(sessionCreateParams.toMap())
    );
```

Reply:
Instead of 
```
    SessionCreateParams sessionCreateParams = SessionCreateParams.builder()
        .putExtraParam("payment_intent_data[application_fee_amount]", 99L)
        .putExtraParam("payment_intent_data[token]", "src_123")
        .build();
```

Reply:
I strongly prefer the latter though. At that point you're already in unsupported territory and it just seems easier, and that's what we go in Go and .NET

Reply:
ah okay.. yup that's reasonable for consistency, and more constrained usage of this extra-params might be better than offering full-blown sub-hash extra params.

cc @ob-stripe if you have more thoughts

Reply:
How hard would it be to support both? I quite like being able to pass maps for nested parameters.

Reply:
If you like that, we should just go with this proposal in that case

Reply:
If we support passing maps with the interface `Map<String, Object>`, we can already support the string key/value with form encoding. So no additional work is required.

One good thing providing this extra params as map value is in the far future when we move to different encoding, we won't be breaking the users here. i.e choice of our lower-level encoding is not exposed to users.

Reply:
👍 Yeah, agreed.